samples/csharp_dotnetcore/49.qnamaker-all-features/Dialog/BotQnAMakerDialog.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.Bot.Builder.AI.QnA.Dialogs;
+using Microsoft.Extensions.Configuration;
+
+namespace Microsoft.BotBuilderSamples.Dialog
+{
+    /// <summary>
+    /// QnAMaker action builder class.  This class exists so that the base QnAMakerDialog can be supplied with
+    /// constructor arguments from configuration.
+    /// </summary>
+    public class BotQnAMakerDialog : QnAMakerDialog
+    {
+        public BotQnAMakerDialog(IConfiguration configuration)
+            :base(configuration["QnAKnowledgebaseId"], configuration["QnAEndpointKey"], BotQnAMakerDialog.GetHostname(configuration["QnAEndpointHostName"]))
+        {
+
+        }
+
+        private static string GetHostname(string hostname)
+        {
+            if (!hostname.StartsWith("http"))
```

Comment:
Instead of mutating, just throw an execption with a clear error message, saying the format of the string is wrong and should follow format XYZ.

Reply:
I chose a safer way to make sure there is a scheme and path, then use Uri to validate, which will throw if not.